_template_python/INDICATOR_DEV_GUIDE.md

```diff
@@ -0,0 +1,446 @@
+* How is the data processed by the data source? E.g. normalization, censoring values with small sample sizes, censoring values associated with low-population areas, smoothing, adding jitter, etc.
+  Keep any code and notes around! They will be helpful for later steps.
+  For any issues that come up, consider now if
+* We’ve seen them before in another dataset and, if so, how we handled it.
+  Is there code around that we can reuse?
+* If it’s a small issue, how would you address it? Do you need an extra function to handle it?
+* If it’s a big issue, talk to others and consider making a PRD to present potential solutions.
+
+
+## Step 2: Pipeline Code
+
+Now that we know the substance and dimensions of our data, we can start planning the pipeline code.
+
+### Logic overview
+
+Broadly speaking, the objective here is to create a script that will download data, transform it (mainly by aggregating it to different geo levels), format it to match our standard format, and save the transformed data to the [receiving directory](https://github.com/cmu-delphi/covidcast-indicators/blob/d36352b/ansible/templates/changehc-params-prod.json.j2#L3) as a CSV.
+The indicator, [validation](https://github.com/cmu-delphi/covidcast-indicators/tree/6912077acba97e835aff7d0cd3d64309a1a9241d/_delphi_utils_python/delphi_utils/validator) (a series of quality checks), and [archive diffing](https://github.com/cmu-delphi/covidcast-indicators/blob/6912077acba97e835aff7d0cd3d64309a1a9241d/_delphi_utils_python/delphi_utils/archive.py) (compressing the data by only outputting rows changed between data versions) are run via the runner.
+Acquisition (ingestion of files from the receiving directory and into the database) is run separately (see the [`delphi-epidata repo`](https://github.com/cmu-delphi/delphi-epidata/tree/c65d8093d9e8fed97b3347e195cc9c40c1a5fcfa)).
+
+`params.json.template` is copied to `params.json` during a run.
+`params.json` is used to set parameters that modify a run and that we expect we’ll want to change in the future e.g. date range to generate) or need to be obfuscated (e.g. API key).
+
+Each indicator includes a makefile (using GNU make), which provides predefined routines for local setup, testing, linting, and running the indicator.
+At the moment, the makefiles use python 3.8.15+.
+
+### Development
+
+To get started, Delphi has a [basic code template](https://github.com/cmu-delphi/covidcast-indicators/tree/6f46f2b4a0cf86137fda5bd58025997647c87b46/_template_python) that you should copy into a top-level directory in the [`covidcast-indicators` repo](https://github.com/cmu-delphi/covidcast-indicators/).
+It can also be helpful to read through other indicators, especially if they share a data source or format.
+
+Indicators should be written in python for speed and maintainability.
+If you think you need to use R, please reconsider! and talk to other engineering team members.
```

Comment:
No, but CTIS (the FB survey) was in R and had many issues with memory and runtime. The backfill projection pipeline is in R but is not yet in production. That might be worth rewriting in Python eventually. It was in R because that's the language Jingjing was using for model development. I don't know how easy or hard it would be to get a Python model to perform exactly the same.

Anyway, do you think I don't need to mention R at all? Just leave it at "Indicators should be written in python..."?

Reply:
Probably don't need to mention it given who's likely developing these pipelines in the future?